src/ray/gcs/gcs_client/service_based_accessor.h

```diff
@@ -9,6 +10,32 @@ namespace gcs {
 
 class ServiceBasedGcsClient;
 
+/// \class Sequencer
+/// Sequencer guarantees that all operations with the same ordering key are sequenced.
+template <class KEY>
+class Sequencer {
+ public:
+  void execute_ordered(KEY key, std::function<void()> operation) {
+    pending_operations_[key].push_back(operation);
+    if (1 == pending_operations_[key].size()) {
+      operation();
+    }
+  }
+
+  void post_execute(KEY key) {
```

Comment:
Consider automatically wrapping the operation in a lambda to call this, to avoid forgetting to call post execute.

Reply:
It is hard to wrap the operation, becuase the post_execute is called in operation callback function.